apps/aemon/src/aemon_mon_on_chain.erl

```diff
@@ -33,70 +52,84 @@ code_change(_FromVsn, St, _Extra) ->
 handle_call(_Req, _From, St) ->
     {reply, {error, unknown_request}, St}.
 
-handle_cast({gen, Height}, St) ->
-    handle_gen(Height, St),
+handle_cast({gen, Height, Type, Hash}, St) ->
+    handle_complete_generation(Height, Type, Hash, St#st.pubkey),
     {noreply, St};
+handle_cast({block, Height, Type, Hash}, St) ->
+    St1 = handle_micro_fork(Height, Type, Hash, St),
+    {noreply, St1};
 handle_cast(_Msg, St) ->
     {noreply, St}.
 
 handle_info(_Msg, St) ->
     {noreply, St}.
 
-%% internals
+%% ==================================================================
+%% internal functions
 
-handle_gen(Height, PubKey) ->
+handle_complete_generation(Height, key, _, PubKey) ->
     try
         {ok, #{micro_blocks := Blocks}} = aec_chain:get_generation_by_height(Height, forward),
+        %% Calculate metrics which require the transaction as input
         [ handle_tx(Height, Tx, aetx_sign:hash(SignTx)) ||
           Block <- Blocks,
           SignTx <- aec_blocks:txs(Block),
           Tx <- [ aetx_sign:tx(SignTx) ],
           aetx:origin(Tx) == PubKey
         ],
-        ttl_gen(Height)
+        %% Forward chain height info to ttl metric worker
+        forward_ttl_gen(Height)
     catch
-        error:badarith -> ok; %% error on payload decode
+        error:badarith ->
+            ok; %% error on payload decode
         _:Reason ->
-            lager:error("~p handle_gen error: ~p", [?MODULE, Reason])
+            lager:error("~p handle_complete_generation error: ~p", [?MODULE, Reason])
     end.
 
+handle_micro_fork(_Height, micro, Hash, #st{micro_blocks = Blocks} = St) ->
+    St#st{micro_blocks = [Hash | Blocks]};
+handle_micro_fork(Height, key, _Hash, #st{micro_blocks = Blocks} = St) ->
+    {ok, #{micro_blocks := BlocksOnChain}} = aec_chain:get_generation_by_height(Height, forward),
+    BlocksOnChain1 = lists:map(
+                       fun(B) ->
+                               {ok, H} = aec_blocks:hash_internal_representation(B),
+                               H
+                       end, BlocksOnChain),
+    %% Check whether we have seen micro-blocks which are not part of the new generation.
+    case [B || B <- Blocks, not lists:member(B, BlocksOnChain1)] of
+        [] ->
+            %% All blocks are in the generation.
+            ok;
+        BlocksNotOnChain ->
+            lager:debug("Monitoring observed micro-fork, missing blocks in generation ~p = ~p",
+                        [Height, BlocksNotOnChain]),
+            ok = aemon_metrics:fork_micro(length(BlocksNotOnChain))
+    end,
+    St#st{micro_blocks = []}.
```

Comment:
So, we collect microblocks and on a key block we zero out them. Also we cross check the generation with our accumulator to see how long is microfork. Looks OK.